---
title: "Plots DS"
output: html_notebook
---



```{r}
library(data.table)
library(ggplot2)
library(ggbeeswarm)
library(viridis)
library(corrplot)
library(tidyr)
library(GenomicRanges)
library(GenomicFeatures)
library(stringr)
```
Get functions
```{r}
source("functions_ds.R")
```
Read gtf
```{r}
  txdb <- makeTxDbFromGFF("C:/Users/kerst/Desktop/MP/Task5/input/splicing_variants.gtf")
```

Read ground truth info
```{r}
info <- read_info("C:/Users/kerst/Desktop/MP/Task5/input/sim_tx_info.txt")
info[type=="template"]$diff_spliced <- F
print("Type distribution:")
table(info$type)
print("Differentially spliced  distribution:")
table(info$diff_spliced)
print("Differentially spliced and type distribution:")
table(info$type,info$diff_spliced)
```
Read ground truth annotation
```{r}
annotation <- read_annotation("C:/Users/kerst/Desktop/MP/Task5/input/event_annotation.tsv")
```

Load introns (because evntpointer uses intronic coordinates :-/ ) (and exons for ir)
```{r}
annotation <- load_introns(annotation,txdb)
```


Read cash outputs
```{r}
cash <- read_ds("C:/Users/kerst/Desktop/MP/Task5/CASH","cash")
cash_counts <- count_ds(cash,"cash",info,annotation,txdb)
print("Type table:")
table(cash$type,cash$real_type)
print("Type table: diff spliced")
table(cash[gene_level==T]$type,cash[gene_level==T]$real_type)
print("Type table: location correct")
table(cash[location_level==T]$type,cash[location_level==T]$real_type)
```
Read eventpointer outputs
```{r}
source("functions_ds.R")
eventpointer <- read_ds("C:/Users/kerst/Desktop/MP/Task5/EventPointer","eventpointer")
eventpointer_counts <- count_ds(eventpointer,"eventpointer",info,annotation,txdb)
print("Type table:")
table(eventpointer$type,eventpointer$real_type)
print("Type table: diff spliced")
table(eventpointer[gene_level==T]$type,eventpointer[gene_level==T]$real_type)
print("Type table: location correct")
table(eventpointer[location_level==T]$type,eventpointer[location_level==T]$real_type)
```
Read aspli
```{r}
source("functions_ds.R")
aspli<- read_ds("C:/Users/kerst/Desktop/MP/Task5/ASpli_du","aspli")
aspli <- pre_filter_type(aspli,info)
aspli_counts <- count_ds(aspli,"aspli",info,annotation,txdb)
print("Type table")
table(aspli$type,aspli$real_type)
print("Type table: diff spliced")
table(aspli[gene_level==T]$type,aspli[gene_level==T]$real_type)
print("Type table: location correct")
table(aspli[location_level==T]$type,aspli[location_level==T]$real_type)

```
Read majiq
```{r}
source("functions_ds.R")
majiq<- read_ds("C:/Users/kerst/Desktop/MP/Task5/majiq_test_10mln_3vs3_deltapsi","majiq")
majiq <- pre_filter_type(majiq,info)
majiq_counts <- count_ds(majiq,"majiq",info,annotation,txdb)
table(majiq$type,majiq$real_type)
print("Type table: diff spliced")
table(majiq[gene_level==T]$type,majiq[gene_level==T]$real_type)
print("Type table: location correct")
table(majiq[location_level==T]$type,majiq[location_level==T]$real_type)

```

Read spladder
```{r}
source("functions_ds.R")
spladder<- read_ds("C:/Users/kerst/Desktop/MP/Task5/spladder","spladder")
spladder<- pre_filter_type(spladder,info)
spladder_counts<- count_ds(spladder,"spladder",info,annotation,txdb)
print("Type table")
table(spladder$type,spladder$real_type)
print("Type table: diff spliced")
table(spladder[gene_level==T]$type,spladder[gene_level==T]$real_type)
print("Type table: location correct")
table(spladder[location_level==T]$type,spladder[location_level==T]$real_type)
```

Read edger
```{r}
source("functions_ds.R")
edger <- read_ds("C:/Users/kerst/Desktop/MP/Task5/edgeR","edger")
edger_counts <- count_ds(edger,"edger",info)
```

Read PSI-SIGMA
```{r}
source("functions_ds.R")
psisigma <- read_ds("C:/Users/kerst/Desktop/MP/Task5/PSI-SIGMA","psisigma")
psisigma <- pre_filter_type(psisigma,info)
psisigma_counts <- count_ds(psisigma,"psisigma",info,annotation,txdb)
print("Type table")
table(psisigma$type,psisigma$real_type)
print("Type table: diff spliced")
table(psisigma[gene_level==T]$type,psisigma[gene_level==T]$real_type)
print("Type table: location correct")
table(psisigma[location_level==T]$type,psisigma[location_level==T]$real_type)

```
Read junctinseq
```{r}
source("functions_ds.R")
junctionseq <- read_ds("C:/Users/kerst/Desktop/MP/Task5/Junction_Seq","junctionseq")
junctionseq_counts <- count_ds(junctionseq,"junctionseq",info)
```


Read dexseq
```{r}
source("functions_ds.R")
dexseq <- read_ds("C:/Users/kerst/Desktop/MP/Task5/DEXseq/dexseq_test_10mln_3vs3","dexseq")
```



Read jsplice
```{r}
source("functions_ds.R")
jsplice <- read_ds("C:/Users/kerst/Desktop/MP/Task5/jSplice","jsplice")
table(jsplice$type)
```
Combine results
```{r}
combined <- rbindlist(list("EdgeR" = edger_counts,
                           "Majiq"=majiq_counts,
                           "Cash"=cash_counts,
                           "ASpli"=aspli_counts,
                           "EventPointer"=eventpointer_counts,
                           "SplAdder"=spladder_counts,
                           "PSI-SIGMA"=psisigma_counts,
                           "JunctionSeq"=junctionseq_counts),idcol="Tool")
```
Plot results
```{r}
combined$level <- factor(sapply(combined$level,str_to_title), levels=c('Gene','Type','Location'))
ggplot(combined,aes(x=tp/(tp+fp),y=tp/(tp+fn),color=Tool))+
  geom_point(size = 3)+
  xlim(0,1)+
  ylim(0,1)+
  facet_grid(~level)+
  coord_fixed()+
  labs(title = "Multi level comparison of differential splicing tools",
       x="Precision",
       y="Recall")+
  theme_bw()+
   theme(
   strip.background = element_rect(
     color="white", fill="white",linetype="blank"
     )
   )
```

Simulate outputs
```{r}
pool <- short_data[diff_spliced==T][1:1000]$transcript_id
tool_names <- paste0("tool_",1:10)
results <- lapply(tool_names,function(x){
  return(sample(pool,500))
})
names(results)<-tool_names
```
Calculate overlaps
```{r}
overlaps <- function(sets){
  matrix <- sapply(sets, function(x){
    vector <- sapply(sets,function(y){
      return(length(intersect(x,y))/length(union(x,y)))
    })
  return(vector)  
  })
  colnames(matrix)<-names(sets)
  rownames(matrix)<-names(sets)

  return(matrix)
}
```
```{r}
matrix <- overlaps(results)
```

```{r}
col <- colorRampPalette(c("#BB4444", "#EE9988", "#FFFFFF", "#77AADD", "#4477AA"))
corrplot(matrix,
         method="square",
         shade.col=NA, 
         tl.col="black", 
         tl.srt=45,
         col=col(200), 
         addCoef.col="black", 
         order="AOE", 
         cl.lim = c(0, 1)
         )
```

